src/Analyzers/Analyzers/test/StartupAnalyzerTest.cs

```diff
@@ -4,347 +4,527 @@
         {
-            // Arrange
-            var source = Read("MvcOptions_UseMvcWithDefaultRouteAndEndpointRoutingDisabled");
+            var source = @"// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-            // Act
-            var diagnostics = await Runner.GetDiagnosticsAsync(source.Source);
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.DependencyInjection;
 
-            // Assert
-            var optionsAnalysis = Assert.Single(Analyses.OfType<OptionsAnalysis>());
-            Assert.True(OptionsFacts.IsEndpointRoutingExplicitlyDisabled(optionsAnalysis));
+namespace Microsoft.AspNetCore.Analyzers.TestFiles.StartupAnalyzerTest
+{
+    public class MvcOptions_UseMvc
+    {
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddMvc();
+        }
 
-            var middlewareAnalysis = Assert.Single(Analyses.OfType<MiddlewareAnalysis>());
-            var middleware = Assert.Single(middlewareAnalysis.Middleware);
-            Assert.Equal("UseMvcWithDefaultRoute", middleware.UseMethod.Name);
+        public void Configure(IApplicationBuilder app)
+        {
+            {|MVC1005:app.UseMvc()|};
+        }
+    }
+}
+";
 
-            Assert.Empty(diagnostics);
+            await VerifyCS.VerifyCodeFixAsync(source, source);
         }
 
         [Fact]
-        public async Task StartupAnalyzer_MvcOptionsAnalysis_AddMvcOptions_FindsEndpointRoutingDisabled()
+        public async Task StartupAnalyzer_MvcOptionsAnalysis_UseMvcAndConfiguredRoutes_FindsEndpointRoutingEnabled()
         {
-            // Arrange
-            var source = Read("MvcOptions_UseMvcWithDefaultRouteAndAddMvcOptionsEndpointRoutingDisabled");
+            var source = @"// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-            // Act
-            var diagnostics = await Runner.GetDiagnosticsAsync(source.Source);
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.DependencyInjection;
 
-            // Assert
-            var optionsAnalysis = Assert.Single(Analyses.OfType<OptionsAnalysis>());
-            Assert.True(OptionsFacts.IsEndpointRoutingExplicitlyDisabled(optionsAnalysis));
+namespace Microsoft.AspNetCore.Analyzers.TestFiles.StartupAnalyzerTest
+{
+    public class MvcOptions_UseMvcAndConfiguredRoutes
+    {
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddMvc();
+        }
 
-            var middlewareAnalysis = Assert.Single(Analyses.OfType<MiddlewareAnalysis>());
-            var middleware = Assert.Single(middlewareAnalysis.Middleware);
-            Assert.Equal("UseMvcWithDefaultRoute", middleware.UseMethod.Name);
```

Comment:
These tests use to verify a lot more details of what's going on than just whether there was a diagnostic. This seems like a downside.

Reply:
The additional data is not exposed to users. At the end of the day, the only part of the analyzers that matters is:

1. The diagnostics (still verified)
2. The code fix (supported by the test framework, but not offered here)
3. The performance (tested through other means)

Testing internal analyzer state involves a large amount of manual overhead (which may or may not be an accurate representation of what is happening during the IDE and/or command line compilation scenarios), and provides very little regression prevention or ability to help diagnose bugs observed by users. Moving to true I/O testing is a huge productivity win in this space.